lib/time/time.go

```diff
@@ -0,0 +1,462 @@
+// interface.
+func (d Duration) Truth() starlark.Bool { return d != 0 }
+
+// Attr gets a value for a string attribute, implementing dot expression support
+// in starklark. required by starlark.HasAttrs interface.
+func (d Duration) Attr(name string) (starlark.Value, error) {
+	return builtinAttr(d, name, durationMethods)
+}
+
+// AttrNames lists available dot expression strings. required by
+// starlark.HasAttrs interface.
+func (d Duration) AttrNames() []string { return builtinAttrNames(durationMethods) }
+
+// Binary implements binary operators, which satisfies the starlark.HasBinary
+// interface.
+func (d Duration) Binary(op syntax.Token, yV starlark.Value, _ starlark.Side) (starlark.Value, error) {
+	x := time.Duration(d)
+	var y time.Duration
+	switch yV.(type) {
+	case starlark.Int:
+		i, ok := yV.(starlark.Int).Int64()
+		if !ok {
+			return nil, fmt.Errorf("duration binary operation: couldn't parse int")
+		}
+		y = time.Duration(i)
+	case Duration:
+		y = time.Duration(yV.(Duration))
+	case Time:
+		y := time.Time(yV.(Time))
+		switch op {
+		case syntax.PLUS:
+			// duration + time = time
+			return Time(y.Add(x)), nil
+		case syntax.MINUS:
+			// duration - time = duration
+			return nil, nil
+		}
+	default:
+		return nil, nil
+	}
+
+	switch op {
+	case syntax.PLUS:
+		return Duration(x + y), nil
+	case syntax.MINUS:
+		return Duration(x - y), nil
+	case syntax.SLASH:
+		if int64(y) == 0 {
+			return nil, fmt.Errorf("time.duration division by zero")
+		}
+		return Duration(x / y), nil
+	case syntax.SLASHSLASH:
+		if int64(y) == 0 {
+			return nil, fmt.Errorf("time.duration floored divison by zero")
+		}
+		return Duration(x / y), nil
+	case syntax.STAR:
+		return Duration(x * y), nil
+	}
+
+	return nil, nil
+}
+
+var durationMethods = map[string]builtinMethod{
+	"hours":       durationhours,
+	"minutes":     durationminutes,
+	"seconds":     durationseconds,
+	"nanoseconds": durationnanoseconds,
+}
+
+func durationhours(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	recv := time.Duration(recV.(Duration))
+	return starlark.Float(recv.Hours()), nil
+}
+
+func durationminutes(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	recv := time.Duration(recV.(Duration))
+	return starlark.Float(recv.Minutes()), nil
+}
+
+func durationseconds(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	recv := time.Duration(recV.(Duration))
+	return starlark.Float(recv.Seconds()), nil
+}
+
+func durationnanoseconds(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	recv := time.Duration(recV.(Duration))
+	return starlark.MakeInt64(recv.Nanoseconds()), nil
+}
+
+// Time is a starlark representation of a point in time.
+type Time time.Time
+
+// String implements the Stringer interface.
```

Comment:
Describe what the function does, not what interface it implements; specifically, the format of the resulting string.